docs/tools/builder/index.rst

```diff
@@ -0,0 +1,329 @@
+        pre_state(
+            (sender, "balance", 1),
+            (receiver, "balance", 0),
+        ),
+        expect(
+            networks=["Frontier"],
+            transaction={
+                "to": receiver,
+                "value": 1,
+                "secretKey": sender_key,
+            },
+            post_state=[
+                [sender, "balance", 0],
+                [receiver, "balance", 1],
+            ]
+        )
+    )
+
+.. note:: 
+
+    Note that :func:`~eth.tools.builder.state.setup_filler` returns a
+    dictionary, whereas all of the following functions such as
+    :func:`~eth.tools.builder.state.pre_state`,
+    :func:`~eth.tools.builder.state.expect`, expect to be passed a dictionary
+    as their single argument and return an updated version of the dictionary.
+
+
+.. autofunction:: eth.tools.builder.state.setup_main_filler
+
+    This function kicks off the filler generation process by creating the general filler scaffold with
+    a test name and general information about the testing environment.
+
+    For tests for the main chain, the `environment` parameter is expected to be a dictionary with some
+    or all of the following keys:
+
+    +------------------------+---------------------------------+
+    | key                    | description                     |
+    +========================+=================================+
+    | ``"currentCoinbase"``  | the coinbase address            |
+    +------------------------+---------------------------------+
+    | ``"currentNumber"``    | the block number                |
+    +------------------------+---------------------------------+
+    | ``"previousHash"``     | the hash of the parent block    |
+    +------------------------+---------------------------------+
+    | ``"currentDifficulty"``| the block's difficulty          |
+    +------------------------+---------------------------------+
+    | ``"currentGasLimit"``  | the block's gas limit           |
+    +------------------------+---------------------------------+
+    | ``"currentTimestamp"`` | the timestamp of the block      |
+    +------------------------+---------------------------------+
+
+
+.. autofunction:: eth.tools.builder.state.pre_state
+
+    This function specifies the state prior to the test execution. Multiple invocations don't override
+    the state but extend it instead.
+
+    In general, the elements of `state_definitions` are nested dictionaries of the following form:
+
+    .. code-block:: python
+
+        {
+            address: {
+                "nonce": <account nonce>,
+                "balance": <account balance>,
+                "code": <account code>,
+                "storage": {
+                    <storage slot>: <storage value>
+                }
+            }
+        }
+
+    To avoid unnecessary nesting especially if only few fields per account are specified, the following
+    and similar formats are possible as well:
```

Comment:
This is a lot of different ways to specify the same thing. It will be harder to spot if a line is a bug, or just one of the 5 different ways to do it correctly. Can we narrow it down a bit?

It doesn't seem that much tougher to always have a `dict` even for the short ones, like:

```py
{address, {'balance': acct_balance}}
{address, {'storage': {slot: val}}
```